src/sentry/rules/history/preview.py

```diff
@@ -198,36 +200,67 @@ def get_fired_groups(
 
 
 def get_top_groups(
-    project: Project, start: datetime, end: datetime, condition_activity: GroupActivityMap
+    project: Project,
+    start: datetime,
+    end: datetime,
+    condition_activity: GroupActivityMap,
+    dataset_map: Dict[str, Dataset],
 ) -> GroupActivityMap:
     """
     Filters the activity to contain only groups that have the most events in the past 2 weeks.
 
     Since frequency conditions require one snuba query per groups, we need to limit the number groups we process.
     """
-    if not condition_activity:
-        return condition_activity
-    # TODO: Also check other datasets for top groups
-    groups = raw_query(
-        dataset=Dataset.Events,
-        start=start,
-        end=end,
-        filter_keys={"project_id": [project.id]},
-        conditions=[("group_id", "IN", list(condition_activity.keys()))],
-        aggregations=[("count", "group_id", "groupCount")],
-        groupby=["group_id"],
-        order_by="-groupCount",
-        seleted_columns=["group_id"],
-        limit=FREQUENCY_CONDITION_GROUP_LIMIT,
-    ).get("data", [])
-
-    top_groups = {group["group_id"] for group in groups}
+    datasets = {dataset_map.get(group) for group in condition_activity.keys()}
+
+    # queries each dataset for top x groups and then gets top x overall
+    groups = []
+    for dataset in datasets:
+        if dataset is None:
+            continue
+        group_ids = list(condition_activity.keys())
+        kwargs = {
+            "dataset": dataset,
+            "start": start,
+            "end": end,
+            "filter_keys": {"project_id": [project.id]},
+            "conditions": [("group_id", "IN", group_ids)],
+            "aggregations": [("count", "group_id", "groupCount")],
+            "groupby": ["group_id"],
+            "order_by": "-groupCount",
+            "seleted_columns": ["group_id"],
+            "limit": FREQUENCY_CONDITION_GROUP_LIMIT,
+        }
+        if dataset == Dataset.Transactions:
+            kwargs["conditions"] = [[["hasAny", ["group_ids", ["array", group_ids]]], "=", 1]]
+            kwargs["aggregations"].append(("arrayJoin", ["group_ids"], "group_id"))
+        groups.extend(raw_query(**kwargs).get("data", []))
+
+    k = lambda x: x["groupCount"]
+    sorted_groups = sorted(groups, key=k, reverse=True)
+
+    top_groups = {group["group_id"] for group in sorted_groups[:FREQUENCY_CONDITION_GROUP_LIMIT]}
     top_activity = {
         group: activity for group, activity in condition_activity.items() if group in top_groups
     }
     return top_activity
 
 
+def get_group_dataset(condition_activity: GroupActivityMap) -> Dict[str, Dataset]:
```

Comment:
Are you sure the return type isn't `Dict[int, Dataset]`?

Reply:
Yeah, not sure about user-provided data that grabs rows from the DB:
https://github.com/getsentry/sentry/blob/1523533be476335003da9ce24a1388fde07e7dc0/src/sentry/rules/history/preview.py#L121-L129

But I would try to parse out the `group_id` as an int. I suspect you're not getting any type incompatibility errors from mypy since you're consistently using str for `group_id` and mypy can't infer the type from some of the functions that retrieves the id, but I think the runtime type is actually an int.